### ISSUE TITLE
Requires caching twice when cache="ram" and batch=-1

### ISSUE THREAD

**Issue by dantetemplar** [ultralytics/ultralytics#20748]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

First it cache while deriving good batch size, and secondly before starting training.

![Image](https://github.com/user-attachments/assets/91cc7ab8-d552-462c-8788-9ea1c2ec1837)

I guess will be better to reuse cache, or to skip cache on batch size estimation.

Also you can see I've monkey patch NUM_THREADS, I have 80 cores processor and I was able to speed up cache initialization by increasing NUM_THREADS from 8 to 80, may be it will be good to add argument that will be passed to BaseDataloader, but it is uses global variable for now: https://github.com/ultralytics/ultralytics/blob/3b818b32ece7f4545a2cb8ed5adc38b31ade26a9/ultralytics/data/base.py#L265.

My full code:
```python
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "ultralytics",
#     "torch",
#     "torchvision",
# ]
# ///

import os
import tempfile

from ultralytics import YOLO

current_full_path = os.path.abspath(__file__)
current_dir = os.path.dirname(current_full_path)
YAML = f"""
path: {current_dir}/HumanRescue
train:
  - 01_train/
  - 02_train/
  - lacmus_train/
val:
  - 03_validation-private/
  - 03_validation-public/
  - lacmus_validation/
test:
  - 04_test/
names:
  0: person
"""

print(YAML)

model = YOLO()


with tempfile.NamedTemporaryFile(delete=True, suffix=".yaml") as f:
    f.write(YAML.encode("utf-8"))
    f.flush()
    f.seek(0)

    data_path = f.name

    import ultralytics.data.base

    ultralytics.data.base.NUM_THREADS = 80
    results = model.train(data=data_path, batch=-1, workers=8, cache="ram")
    # , deterministic=False (to avoid warning)
    # batch=96 (to avoid caching twice)
    print(results)
```

### Environment

```
Ultralytics 8.3.141 🚀 Python-3.12.9 torch-2.7.0+cu126 CUDA:0 (Tesla V100-SXM2-32GB, 32501MiB)
Setup complete ✅ (80 CPUs, 503.8 GB RAM, 245.1/438.1 GB disk)

OS                  Linux-4.15.0-161-generic-x86_64-with-glibc2.31
Environment         Docker
Python              3.12.9
Install             git
Path                /home/robo/resciu-dataset/.venv/lib/python3.12/site-packages/ultralytics
RAM                 503.79 GB
Disk                245.1/438.1 GB
CPU                 Intel Xeon E5-2698 v4 2.20GHz
CPU count           80
GPU                 Tesla V100-SXM2-32GB, 32501MiB
GPU count           1
CUDA                12.6

numpy               ✅ 2.2.4>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.7.0>=1.8.0
torch               ✅ 2.7.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.22.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example

```python
model.train(data="visdrone.yaml", batch=-1, workers=8, cache="ram")
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @dantetemplar, thank you for your detailed report and suggestions for Ultralytics 🚀! We appreciate your insights, especially regarding cache reuse and NUM_THREADS optimization.

This is an automated response to help you get started. An Ultralytics engineer will review your issue and assist you soon.

If this is a 🐛 Bug Report, thank you for including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/)! This helps us debug issues more efficiently.

If you have further details or logs, please add them here. For new users, we recommend visiting the [Docs](https://docs.ultralytics.com/) for useful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, as well as the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- Or join our [Subreddit](https://reddit.com/r/Ultralytics) to share and learn from others

## Upgrade

To ensure your issue persists with the latest release, please upgrade the `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified, up-to-date environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for helping improve Ultralytics!

**Comment by glenn-jocher** [comment]
Thanks for reporting this issue @dantetemplar! You've identified a valid optimization opportunity. When using `batch=-1` with `cache="ram"`, the system does indeed cache images twice - first during batch size estimation, then again for the actual training.

This happens because the `auto_batch` method in `DetectionTrainer` builds a dataset during estimation without considering the cache status. The suggested approaches are good:

1. Skip caching during batch size estimation (preferred)
2. Reuse the cache from the estimation phase during training

Your observation about `NUM_THREADS` is also valuable. Adding a configurable parameter to control thread count for caching would be beneficial for systems with many cores like yours.

I'll create an issue to track these improvements. As a workaround, you can either:
- Explicitly set the batch size to avoid the auto-determination
- Use `cache="disk"` instead which is more memory-efficient (though slower)

Would you be interested in contributing a PR to fix this? It would likely involve modifying the `auto_batch` method to either skip caching or preserve the cache for reuse.

**Comment by XBastille** [comment]
Hi @glenn-jocher , I would like to work on this issue, I am thinking about skipping caching during batch size estimation and Add NUM_THREADS configuration (as separate smaller feature)

**Comment by XBastille** [comment]
Hi @glenn-jocher I have opened a pr #20891 , kindly check!!! 🙏🏼

**Comment by glenn-jocher** [comment]
Thanks for taking this on @XBastille! I'll review PR #20891 and provide feedback there. Great to see community contributions addressing both the double caching issue and the NUM_THREADS configurability that @dantetemplar identified.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR (#20891). The PR focuses on optimizing cache management during auto-batch training by introducing a dedicated context manager for configuration overrides and improving memory handling. Key improvements include:

- Cleaner and more maintainable code by moving patching utilities to a centralized location.
- Safer temporary changes to training configurations, ensuring they don't persist unexpectedly.
- More efficient memory usage by explicitly releasing temporary datasets.
- Enhanced reliability for ONNX export workflows.

Thanks to XBastille for authoring the PR, with valuable contributions from dantetemplar, glenn-jocher, Laughing-q, and Y-T-G.

To test if these changes resolve your issue, try installing the latest version directly from the Ultralytics GitHub repository:

```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release which will incorporate these updates.

Please let us know if this fix addresses your issue, or if you're still experiencing any problems. Thank you for reporting this, and feel free to share any additional feedback! 🙏
